src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -360,6 +362,182 @@ bool emberAfOperationalCredentialsClusterAddOpCertCallback(chip::app::Command *
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(3), certificationElement)); // TODO: add client_category_id
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(4), certificationElement)); // TODO: add security_level
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(5), certificationElement)); // TODO: add security_information
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(6), certificationElement)); // TODO: add version_number
+    SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
+    SuccessOrExit(err = tlvWriter.Finalize(&certDeclarationTbs));
+
+    SuccessOrExit(err = keypair.ECDSA_sign_msg(certDeclarationTbs->Start(), certDeclarationTbs->DataLength(), signature));
+
+    certDeclaration = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    VerifyOrExit(!certDeclaration.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    tlvWriter.Init(std::move(certDeclaration));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(0), ByteSpan(certDeclarationTbs->Start(), certDeclarationTbs->DataLength())));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(1), ByteSpan(signature, signature.Length())));
+    SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
+    SuccessOrExit(err = tlvWriter.Finalize(&certDeclaration));
+
+    firmwareInfo = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    VerifyOrExit(!firmwareInfo.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    tlvWriter.Init(std::move(firmwareInfo));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(0), certificationElement));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(1), ByteSpan(nullptr, 0))); // TODO: add firmware digest(s)
+    SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
+    SuccessOrExit(err = tlvWriter.Finalize(&firmwareInfo));
+
+    attestationElements = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    VerifyOrExit(!attestationElements.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    attestationElementsTbs = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    VerifyOrExit(!attestationElementsTbs.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    tlvWriter.Init(std::move(attestationElements));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(0), ByteSpan(certDeclaration->Start(), certDeclaration->DataLength())));
+    SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(1), attestationNonce));
```

Comment:
attestationNonce is used without validating that its length is exactly 32

Reply:
Updated